File: backend/app/models/product.py
"""
产品模型
"""
from sqlalchemy import Column, Integer, String, Date, DateTime, Enum, ForeignKey, Boolean, Text
from sqlalchemy.types import Numeric
from sqlalchemy.sql import func
from sqlalchemy.orm import relationship
import enum

from app.core.database import Base


class ProductStatus(str, enum.Enum):
    """产品状态"""
    ACTIVE = "active"  # 运行中
    LIQUIDATED = "liquidated"  # 已清盘
    SUSPENDED = "suspended"  # 暂停


class Product(Base):
    """产品表"""
    __tablename__ = "products"
    
    id = Column(Integer, primary_key=True, index=True, comment="产品ID")
    product_code = Column(String(50), unique=True, nullable=False, index=True, comment="产品代码")
    product_name = Column(String(200), nullable=False, comment="产品名称")
    manager_id = Column(Integer, ForeignKey("managers.id"), nullable=False, comment="管理人ID")
    strategy_type = Column(String(50), comment="策略类型")
    established_date = Column(Date, comment="成立日期")
    liquidation_date = Column(Date, comment="清盘日期")
    management_fee = Column(Numeric(5, 2), comment="管理费率(%)")
    performance_fee = Column(Numeric(5, 2), comment="业绩报酬(%)")
    benchmark_code = Column(String(50), comment="基准代码")
    benchmark_name = Column(String(100), comment="基准名称")
    is_invested = Column(Boolean, default=False, comment="是否在投")
    status = Column(String(20), default="active", comment="状态")
    remark = Column(String(1000), comment="备注")
    analysis_data = Column(Text, nullable=True, comment="分析数据（回撤、胜率、滚动分析等）JSON格式")
    attribution_data = Column(Text, nullable=True, comment="风格归因数据JSON格式")
    last_analysis_update = Column(DateTime, nullable=True, comment="分析数据最后更新时间")
    created_at = Column(DateTime, server_default=func.now(), comment="创建时间")
    updated_at = Column(DateTime, server_default=func.now(), onupdate=func.now(), comment="更新时间")
    
    # 关联关系
    manager = relationship("Manager", back_populates="products")
    nav_data = relationship("NavData", back_populates="product")
    
    def __repr__(self):
        return f"<Product(id={self.id}, code='{self.product_code}', name='{self.product_name}')>"

File: frontend/src/types/product.ts
/**
 * 产品相关类型定义
 */

import type { PaginationParams, PaginationResponse, SearchParams } from './api'

// 产品状态
export type ProductStatus = 'active' | 'liquidated' | 'suspended'

// 产品基础信息
export interface Product {
  id: number
  product_code: string
  product_name: string
  manager_id: number
  manager_name?: string
  strategy_type?: string
  established_date?: string
  liquidation_date?: string
  management_fee?: number
  performance_fee?: number
  benchmark_code?: string
  benchmark_name?: string
  is_invested?: boolean
  status: ProductStatus
  remark?: string
  created_at: string
  updated_at: string
  // 兼容别名
  code?: string
  name?: string
}

// 创建产品请求
export interface ProductCreate {
  product_code: string
  product_name: string
  manager_id: number
  strategy_type?: string
  established_date?: string
  liquidation_date?: string
  management_fee?: number
  performance_fee?: number
  benchmark_code?: string
  benchmark_name?: string
  is_invested?: boolean
  status?: ProductStatus
  remark?: string
}

// 更新产品请求
export interface ProductUpdate {
  product_code?: string
  product_name?: string
  manager_id?: number
  strategy_type?: string
  established_date?: string
  liquidation_date?: string
  management_fee?: number
  performance_fee?: number
  benchmark_code?: string
  benchmark_name?: string
  is_invested?: boolean
  status?: ProductStatus
  remark?: string
}

// 产品列表查询参数
export interface ProductListParams extends PaginationParams, SearchParams {
  manager_id?: number
  strategy_type?: string
  status?: string
}

// 产品列表响应
export interface ProductListResponse extends PaginationResponse<Product> {}

// 产品统计信息
export interface ProductStatistics {
  total: number
  by_status: Record<string, number>
  by_strategy: Record<string, number>
}

// 产品状态选项
export const PRODUCT_STATUS_OPTIONS = [
  { label: '运行中', value: 'active' },
  { label: '已清盘', value: 'liquidated' },
  { label: '暂停', value: 'suspended' }
] as const

// 产品状态颜色映射
export const PRODUCT_STATUS_COLOR: Record<ProductStatus, string> = {
  'active': 'success',
  'liquidated': 'danger',
  'suspended': 'warning'
}

// 产品状态中文映射
export const PRODUCT_STATUS_TEXT: Record<ProductStatus, string> = {
  'active': '运行中',
  'liquidated': '已清盘',
  'suspended': '暂停'
}
